Extract more meta data from output

LJ1102 opened the pull request:
Metadata is being parsed on a per stream basis. This actually removes the `meta` event and instead provides the metadata as third parameter to the `exit` event.

The resulting metadata(ie. pokemon_card.flv -> pokemon_card.webm) object looks like this:

```javascript
{
    input: {
        duration: 32056,
        start: 0,
        bitrate: null,
        stream: [
            [
                {
                    type: "video",
                    codec: "h264",
                    format: "yuv420p",
                    resolution: [ 320, 240 ],
                    bitrate: 202,
                    fps: 29.92
                },{
                    type: "audio",
                    codec: "aac",
                    samplerate: 22050,
                    channels: 2,
                    encoding: "fltp",
                    bitrate: 63
                }
            ]
        ]
    },
    output: {
        stream: [
            [
                {
                    type: "video",
                    codec: "libvpx",
                    format: "yuv420p",
                    resolution: [ 320, 240 ],
                    bitrate: 200
                },{
                    type: "audio",
                    codec: "libvorbis",
                    samplerate: 22050,
                    channels: 2,
                    encoding: "fltp",
                    bitrate: null
                }
            ]
        ]
    }
}
```

binarykitchen commented on avconv.js:
```diff
@@ -47,27 +47,98 @@ function findTime(data) {
+            if (ioData.duration)
+                ioData.duration = toMilliSeconds(ioData.duration);
+            if (ioData.bitrate)
+                ioData.bitrate = parseInt(ioData.bitrate, 10);
```
reuse mapInts() here

binarykitchen commented on avconv.js:
```diff
@@ -47,27 +47,98 @@ function findTime(data) {
+        streamIndex,
+        streamData,
+        metaType,
+        ioData, tmp;
```
rename `ioData` to avoid confusion with io.js itself

binarykitchen commented on avconv.js:
```diff
@@ -47,27 +47,98 @@ function findTime(data) {
+        metaType,
+        ioData, tmp;
+
+    function mapInts(v) {return parseInt(v,10)}
```
function `mapInts()` does not map anything itself, so change it to `formatInts()` or so.

binarykitchen commented on avconv.js:
```diff
@@ -47,27 +47,98 @@ function findTime(data) {
+                streamData.codec = tmp[VIDEO_META.CODEC];
+                streamData.format = tmp[VIDEO_META.FORMAT];
+                streamData.resolution = tmp[VIDEO_META.RESOLUTION].split("x").map(mapInts);
+                streamData.bitrate = parseInt(tmp[
```
mapInts() too

binarykitchen commented on avconv.js:
```diff
@@ -47,27 +47,98 @@ function findTime(data) {
+                    streamData.fps = parseFloat(tmp[VIDEO_META.FPS]);
+            } else if (streamData.type == "audio") {
+                streamData.codec = tmp[AUDIO_META.CODEC];
+                streamData.samplerate = parseInt(tmp[AUDIO_META.SAMPLERATE], 10);
```
mapInts() too

binarykitchen commented on avconv.js:
```diff
@@ -47,27 +47,98 @@ function findTime(data) {
+                streamData.samplerate = parseInt(tmp[AUDIO_META.SAMPLERATE], 10);
+                streamData.channels = getChannelCount(tmp[AUDIO_META.SPATIALIZATION]);
+                streamData.encoding = tmp[AUDIO_META.ENCODING];
+                streamData.bitrate = parseInt(tmp[AUDIO_META.BITRATE], 10);
```
mapInts() too

binarykitchen commented on avconv.js:
```diff
@@ -160,7 +221,7 @@ module.exports = function avconv(params) {
     avconv.on(eventType, function(exitCode, signal) {
         stream.end();
-        stream.emit('exit', exitCode, signal);
+        stream.emit('exit', exitCode, signal, parseMetaData(output));
```
what if i am not interested in meta data in my app? can this be suppressed to gain speed?

LJ1102 replied:
Well, we could reintroduce the `meta` event and use `stream.listeners('meta')` to check if there are listeners for it, but compared to the time avconv takes to do its job optimizing this seems unnecessary.

binarykitchen replied:
well, you are right, but i have a time sensitive app here using this package.

tell me how long it takes to parse meta data?

LJ1102 replied:
benchmarks with 1000 iterations averaged around 0.06 ms on a 1.9 GHz machine under medium load so spawning avconv probably takes longer than parsing its meta data ;)

binarykitchen replied:
good to know. alright, leave it there then. thanks for benchmarking!

binarykitchen commented:
wonderful, thanks heaps man!

good code but where is the test code?

furthermore i commented right in your code. can you do these changes and do another PR?

LJ1102 commented:
I've adjusted the unit tests, and renamed the variables/functions, lets see what travis says. Pushing them into the branch should be enough, no need for another PR or is there?

binarykitchen commented:
perfect, i am about to merge your PR ... but oh, i am sorry i forgot to ask to update the README.md as well. these chapters are outdated now:

- https://github.com/binarykitchen/avconv#how-to-watch-for-results-progress-meta-data-output-errors-exit-code
- https://github.com/binarykitchen/avconv#stream--avconvparams
- and probably https://github.com/binarykitchen/avconv/blob/master/History.md#200--2014-05-14 too ... this will be breaking, so make it version 3

what do you think? sorry and thanks for your patience!

LJ1102 commented:
My last commit updates package.json, readme.md aswell as history.md. I've added myself to the contributors if thats ok for you, if not just say it and I'll `fixup` the commit. Feel free to rewrite the readme stuff as I'm not really good at writing human readable texts ;)

binarykitchen commented:
wonderful, i will make minor adjustments tonight at home and publish to npm

but i am afraid, tests fail on travis:

```
✖ TC 3: do a conversion - convert pokemon flv to webm
Assertion Message: Input audio stream encoding is correct
AssertionError: 's16' === 'fltp'
    at Object.strictEqual (/home/travis/build/binarykitchen/avconv/node_modules/nodeunit/lib/types.js:83:39)
    at AvStream.<anonymous> (/home/travis/build/binarykitchen/avconv/tests/basics.js:213:19)
    at AvStream.g (events.js:180:16)
    at AvStream.emit (events.js:106:17)
    at ChildProcess.<anonymous> (/home/travis/build/binarykitchen/avconv/avconv.js:223:16)
    at ChildProcess.emit (events.js:98:17)
    at maybeClose (child_process.js:766:16)
    at Socket.<anonymous> (child_process.js:979:11)
    at Socket.emit (events.js:95:17)
    at Pipe.close (net.js:466:12)
Assertion Message: Output audio stream encoding is correct
AssertionError: 's16' === 'fltp'
    at Object.strictEqual (/home/travis/build/binarykitchen/avconv/node_modules/nodeunit/lib/types.js:83:39)
    at AvStream.<anonymous> (/home/travis/build/binarykitchen/avconv/tests/basics.js:233:19)
    at AvStream.g (events.js:180:16)
    at AvStream.emit (events.js:106:17)
    at ChildProcess.<anonymous> (/home/travis/build/binarykitchen/avconv/avconv.js:223:16)
    at ChildProcess.emit (events.js:98:17)
    at maybeClose (child_process.js:766:16)
    at Socket.<anonymous> (child_process.js:979:11)
    at Socket.emit (events.js:95:17)
    at Pipe.close (net.js:466:12)
```

LJ1102 commented:
I digged into this, and it turns out that the `encoding` is the `sampleformat` and depends on the decoder/encoder used, the chosen sampleformat varies from platform to platform so I removed the test for it, also I renamed the metadata member from `encoding` to `sampleformat` to avoid any confusions.

binarykitchen commented:
all looks great, published to npm - again many thanks man!
